src/lib/schedule.c

```diff
@@ -343,7 +343,25 @@ void schedule_task_complete(struct task *task)
 	tracev_pipe("schedule_task_complete()");
 
 	spin_lock_irq(&sch->lock, flags);
-	task->state = TASK_STATE_COMPLETED;
+
+	/* Some high priority HW based IRQ handlers can reschedule tasks
+	 * immediately. i.e. before the task context can change task state
+	 * back to COMPLETED. Check here to make sure we dont clobber
+	 * task->state for regular non IRQ users.
+	 */
+	switch (task->state) {
+	case TASK_STATE_RUNNING:
+		task->state = TASK_STATE_COMPLETED;
+		break;
+	case TASK_STATE_QUEUED:
+		/* nothing to do here, high priority IRQ has scheduled us */
+		break;
+	default:
```

Comment:
@lgirdwood what if the task state changes to pending because of the race condition we saw earlier. We will still end up setting it to completed and miss the next ipc task.

Reply:
Potentially if the schedule time for the next task is quite close and in this case it is immediately :
```
void ipc_schedule_process(struct ipc *ipc)
{
	schedule_task(&ipc->ipc_task, 0, 100);
}
```
We need to relax this as IPC is never that high priority. Added PENDING case and pushed.